notebooks/Threshold Optimization Postprocessing for Binary Classification.ipynb

```diff
@@ -0,0 +1,452 @@
+    "            x_tick_labels_by_label.extend([\"{} recidivism\".format(group), \"{} no recidivism\".format(group)])\n",
+    "    \n",
+    "    x_tick_labels = x_tick_labels_basic + x_tick_labels_by_label\n",
+    "    plt.xticks(range(1, len(x_tick_labels)+1), x_tick_labels, rotation=45, horizontalalignment=\"right\")\n",
+    "        \n",
+    "show_proportions(X_train, sensitive_features_train, y_train, description=\"original training data:\", plot_row_index=1)\n",
+    "show_proportions(X_train, sensitive_features_train, unconstrained_predictor.predict(X_train), y_train, description=\"fairness-unaware prediction on training data:\", plot_row_index=2)\n",
+    "show_proportions(X_test, sensitive_features_test, y_test, description=\"original test data:\", plot_row_index=3)\n",
+    "show_proportions(X_test, sensitive_features_test, unconstrained_predictor.predict(X_test), y_test, description=\"fairness-unaware prediction on test data:\", plot_row_index=4)\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We notice a stark contrast in the predictions with males being a lot more likely to be predicted to get approved, similar to the original training data. However, there's even a disparity between the subgroup of males and females that are approved. When considering only the samples labeled with \"loan not repaid\" males are much more likely to be predicted to get approved than females. The test data shows a similar disparity."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from fairlearn.widget import FairlearnDashboard\n",
+    "\n",
+    "FairlearnDashboard(sensitive_features=sensitive_features_test.values,\n",
+    "                   true_y=y_test.values,\n",
+    "                   predicted_ys=unconstrained_predictor.predict(X_test),\n",
+    "                   class_names=None,\n",
+    "                   feature_names=X_test.columns.values.tolist(),\n",
+    "                   is_classifier=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Postprocessing the model to get a fair model"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The idea behind postprocessing is to alter the output of the fairness-unaware model to achieve fairness. The postprocessing algorithm requires three input arguments:\n",
+    "- the matrix of samples X\n",
+    "- the vector of predictions y from the fairness-unaware model \n",
+    "- the vector of group attribute values A (in our code we refer to it as `sensitive_features`)\n",
+    "\n",
+    "The goal is to make the output fair with respect to constraints. Our postprocessing algorithm uses one of\n",
+    "- Demographic Parity (DP): $P[h(X)=\\hat{y} | A=a] = P[h(X)=\\hat{y}] \\qquad \\forall a, \\hat{y}$\n",
```

Comment:
the white space  around the |  isn't showing. It looks a bit crammed when rendered as markup, is there a latex whitespace char?

Reply:
Indeed. I found that and fixed it.